hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/RegionReplicaInfo.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.hadoop.hbase.HRegionLocation;
+import org.apache.hadoop.hbase.MetaTableAccessor;
+import org.apache.hadoop.hbase.RegionLocations;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.client.RegionInfo;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.master.RegionState;
+import org.apache.hadoop.hbase.master.assignment.RegionStateStore;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * A POJO that consolidates the information about a single region replica that's stored in meta.
+ */
+@InterfaceAudience.Private
+public final class RegionReplicaInfo {
+  private final byte[] row;
+  private final RegionInfo regionInfo;
+  private final byte[] regionName;
+  private final byte[] startKey;
+  private final byte[] endKey;
+  private final Integer replicaId;
+  private final RegionState.State regionState;
+  private final ServerName serverName;
+
+  private RegionReplicaInfo(final Result result, final HRegionLocation location) {
+    final Optional<Result> maybeResult = Optional.ofNullable(result);
+    final Optional<HRegionLocation> maybeLocation = Optional.ofNullable(location);
+    final Optional<RegionInfo> maybeRegionInfo = maybeLocation.map(HRegionLocation::getRegion);
+
+    this.row = maybeResult.map(Result::getRow).orElse(null);
+    this.regionInfo = maybeRegionInfo.orElse(null);
+    this.regionName = maybeRegionInfo.map(RegionInfo::getRegionName).orElse(null);
+    this.startKey = maybeRegionInfo.map(RegionInfo::getStartKey).orElse(null);
+    this.endKey = maybeRegionInfo.map(RegionInfo::getEndKey).orElse(null);
+    this.replicaId = maybeRegionInfo.map(RegionInfo::getReplicaId).orElse(null);
+    this.regionState = result != null && maybeRegionInfo.isPresent()
+      ? RegionStateStore.getRegionState(result, maybeRegionInfo.get())
+      : null;
+    this.serverName = maybeLocation.map(HRegionLocation::getServerName).orElse(null);
+  }
+
+  public static List<RegionReplicaInfo> from(final Result result) {
+    if (result == null) {
+      return Collections.singletonList(null);
+    }
+
+    final RegionLocations locations = MetaTableAccessor.getRegionLocations(result);
+    if (locations == null) {
+      return Collections.singletonList(null);
+    }
+
+    return StreamSupport.stream(locations.spliterator(), false)
+      .map(location -> new RegionReplicaInfo(result, location))
+      .collect(Collectors.toList());
+  }
+
+  public byte[] getRow() {
+    return row;
+  }
+
+  public RegionInfo getRegionInfo() {
+    return regionInfo;
+  }
+
+  public byte[] getRegionName() {
+    return regionName;
+  }
+
+  public byte[] getStartKey() {
+    return startKey;
+  }
+
+  public byte[] getEndKey() {
+    return endKey;
+  }
+
+  public Integer getReplicaId() {
+    return replicaId;
+  }
+
+  public RegionState.State getRegionState() {
+    return regionState;
+  }
+
+  public ServerName getServerName() {
```

Comment:
These have to be public?

Reply:
For the same reason as the accessors in `MetaBrowser`: they need to be reachable by `table.jsp`.